modules/v2x/v2x_proxy/obu_interface/grpc_interface/grpc_server.cc

```diff
@@ -21,54 +21,70 @@
 
 #include "modules/v2x/v2x_proxy/obu_interface/grpc_interface/grpc_server.h"
 
+#include <chrono>
+
+#include "modules/v2x/common/v2x_proxy_gflags.h"
+
 namespace apollo {
 namespace v2x {
 
-using apollo::perception::PerceptionObstacles;
+using ::apollo::v2x::StatusResponse;
+using ::apollo::v2x::obu::OBURsi;
+using ::apollo::v2x::obu::OBUTrafficLightData;
 using grpc::Status;
 
-GrpcServerImpl::GrpcServerImpl() : node_(cyber::CreateNode("v2x_grpc_server")) {
-  ACHECK(node_) << "Create v2x grpc server node failed";
-  first_flag_writer_ =
-      node_->CreateWriter<StatusResponse>("/apollo/v2x/inner/sync_flag");
-  ACHECK(first_flag_writer_) << "Create sync flag writer failed";
+GrpcServerImpl::GrpcServerImpl()
+    : node_(::apollo::cyber::CreateNode("v2x_grpc_server")) {
+  CHECK(!!node_);
+  // _monitor = cybertron::MetricController::Instance();
   AINFO << "GrpcServerImpl initial success";
   init_flag_ = true;
 }
 
-grpc::Status GrpcServerImpl::SendPerceptionObstacles(
-    grpc::ServerContext* /* context */, const PerceptionObstacles* request,
-    StatusResponse* response) {
-  ADEBUG << "Received SendPerceptionObstacles request from client! \n";
-  if (request->perception_obstacle().empty()) {
-    response->set_status(false);
-    response->set_info("error perception obstacle size == 0");
-    AERROR << "SendPerceptionObstacles request has no perception obstacle";
-    return Status::CANCELLED;
-  }
+grpc::Status GrpcServerImpl::SendV2xRSI(grpc::ServerContext * /* context */,
+                                        const OBURsi *request,
+                                        StatusResponse *response) {
+  AINFO << "Received SendV2xRSI request from client! \n";
+  AINFO << request->DebugString();
+
   if (!request->has_header()) {
     response->set_status(false);
-    response->set_info("error no header in PerceptionObstacles request");
-    AERROR << "SendPerceptionObstacles request has no header";
+    response->set_info("error no header in RSI request");
+    AERROR << "SendV2xRSI request has no header";
     return Status::CANCELLED;
   }
   {
-    std::lock_guard<std::mutex> guard(obstacles_mutex_);
-    latest_obstacles_.CopyFrom(*request);
+    std::lock_guard<std::mutex> guard(rsi_mutex_);
+    latest_rsi_.CopyFrom(*request);
+    auto obu_header = latest_rsi_.mutable_header();
+    if (obu_header->has_timestamp_sec()) {
+      AINFO << "OBU Timestamp: " << std::setiosflags(std::ios::fixed)
+            << std::setprecision(11) << obu_header->timestamp_sec();
+      AINFO << "SVB Timestamp: " << std::setiosflags(std::ios::fixed)
+            << std::setprecision(11) << ::apollo::cyber::Time::Now().ToSecond();
+      obu_header->set_timestamp_sec(::apollo::cyber::Time::Now().ToSecond());
+    }
+    rsi_refresh_ = true;
```

Comment:
Shouldn't rsi_refresh protected by the lock ?

Reply:
It has beed protected by ```rsi_mutex_```

Reply:
> It has beed protected by `rsi_mutex_`

The lock unlocks at L66.

Reply:
OK. I mistook the bracket.